### PR TITLE
chore: update to Go 1.21

### PR DESCRIPTION
**Pull request by jamestelfer**
# What does this change

Updates build to use Go 1.21, and tries to ensure CI workflows are operating against the version of Go currently used by `go.mod`.

Go 1.21 includes various enhancements including new APIs for `slice`, along with a variety of fixes.

# What issue does it fix

No specific issue, just looking to be able to use some newer APIs.

# Notes for the reviewer

Some of the workflows only execute on `master`. The changes appear safe, and any issues could potentially be fixed in a "fast follow" response.
# Checklist
- [x] Made sure the PR follows the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines
- [x] Tests if something new is added

### CODE REVIEW

File: .github/workflows/linting.yaml [hunk 1]
```diff
@@ -10,4 +10,5 @@ jobs:
       - name: golangci-lint
         uses: reviewdog/action-golangci-lint@v2
         with:
+          go_version_file: "go.mod"
```
Comment (jamestelfer) [inline .github/workflows/linting.yaml:13]: Note: reviewdog uses snake case for its arguments.

File: go.sum
Comment (jamestelfer) [inline go.sum:1]: Gonna be honest, don't really understand these changes.

Comment (lindell) [reply go.sum:1]: Might change between go versions, but should be backwards and forwards compatible.

**Review by lindell — APPROVED** [review]
LGTM 👍 

Didn't know about the go-version-file, makes it easier to have versioned updated of the go toolchain. Even though it's generally not a problem 😄

**Pull request merged**